src/passes/TypeMerging.cpp

```diff
@@ -571,77 +610,104 @@ void TypeMerging::applyMerges() {
-    return false;
-  }
-  auto aKind = a.getKind();
-  auto bKind = b.getKind();
-  if (aKind != bKind) {
-    return false;
+  // in the DFA, i.e. any children that are not nontrivial references. We treat
+  // full descriptor chains as single units, so compare the shape of every type
+  // in the chains rooted at `a` and `b`.
+  assert(!a.getDescribedType() && !b.getDescribedType());
+  auto chainA = a.getDescriptorChain();
+  auto chainB = b.getDescriptorChain();
+  auto itA = chainA.begin();
+  auto itB = chainB.begin();
+  while (itA != chainA.end() && itB != chainB.end()) {
+    a = *itA++;
+    b = *itB++;
+    if (a.isOpen() != b.isOpen()) {
+      return false;
+    }
+    if (a.isShared() != b.isShared()) {
+      return false;
+    }
+    // Ignore supertype because we want to be able to merge into parents.
+    auto aKind = a.getKind();
+    auto bKind = b.getKind();
+    if (aKind != bKind) {
+      return false;
+    }
+    switch (aKind) {
+      case HeapTypeKind::Func:
+        if (!shapeEq(a.getSignature(), b.getSignature())) {
+          return false;
+        }
+        break;
+      case HeapTypeKind::Struct:
+        if (!shapeEq(a.getStruct(), b.getStruct())) {
+          return false;
+        }
+        break;
+      case HeapTypeKind::Array:
+        if (!shapeEq(a.getArray(), b.getArray())) {
+          return false;
+        }
+        break;
+      case HeapTypeKind::Cont:
+        WASM_UNREACHABLE("TODO: cont");
+      case HeapTypeKind::Basic:
+        WASM_UNREACHABLE("unexpected kind");
+    }
   }
-  switch (aKind) {
-    case HeapTypeKind::Func:
-      return shapeEq(a.getSignature(), b.getSignature());
-    case HeapTypeKind::Struct:
-      return shapeEq(a.getStruct(), b.getStruct());
-    case HeapTypeKind::Array:
-      return shapeEq(a.getArray(), b.getArray());
-    case HeapTypeKind::Cont:
-      WASM_UNREACHABLE("TODO: cont");
-    case HeapTypeKind::Basic:
-      WASM_UNREACHABLE("unexpected kind");
-  }
-  return false;
+  return itA == chainA.end() && itB == chainB.end();
 }
 
 size_t shapeHash(HeapType a) {
-  size_t digest = hash(a.isOpen());
-  rehash(digest, a.isShared());
-  // Ignore supertype because we want to be able to merge into parents.
-  rehash(digest, !!a.getDescriptorType());
-  rehash(digest, !!a.getDescribedType());
-  auto kind = a.getKind();
-  rehash(digest, kind);
-  switch (kind) {
-    case HeapTypeKind::Func:
-      hash_combine(digest, shapeHash(a.getSignature()));
-      return digest;
-    case HeapTypeKind::Struct:
-      hash_combine(digest, shapeHash(a.getStruct()));
-      return digest;
-    case HeapTypeKind::Array:
-      hash_combine(digest, shapeHash(a.getArray()));
-      return digest;
-    case HeapTypeKind::Cont:
-      WASM_UNREACHABLE("TODO: cont");
-    case HeapTypeKind::Basic:
-      break;
+  assert(!a.getDescribedType());
+  size_t digest = 0xA76F35EC;
+  for (auto type : a.getDescriptorChain()) {
+    rehash(digest, 0xCC6B0DD9);
```

Comment:
What are these curious-looking constants?

Reply:
I googled for "random number between 0 and 4294967295" :)

Reply:
I see...

I think in general we just use 1, 2, 3 etc. here, keeping them different from each other.

Reply:
There are some places where we do that and other places where we use hard coded random (looking) numbers. Generally using the large random numbers for control flow seems useful to prevent clashes with small bools or enum values.

Reply:
Fair enough, sgtm.